### PR TITLE
External SD2

### PR DESCRIPTION
**Pull request by przybylski**
Implement mechanism with @tobiasKaminsky to allow users to specify their owncloud storage path.
Together with migration mechanism.
https://github.com/owncloud/android/issues/10

### CODE REVIEW

**Comment by tobiasKaminsky** [comment]
Merged in beta :)

File: src/com/owncloud/android/utils/DataStorageUtils.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+		String storagePath = Environment.getExternalStorageDirectory().getPath();
+		if (canBeAddedToAvailableList(storagePath))
+			list.add(new Storage(context.getString(R.string.storage_description_external),
+					 			 storagePath));
+
+		storagePath = System.getenv("SECONDARY_STORAGE"); // : separated paths to sd cards
+		list.addAll(getSDCardStorage(storagePath, context));
+
+		list.add(new Storage("Costam", Environment.getExternalStorageDirectory().getAbsolutePath() + "/costam"));
```
Comment (tobiasKaminsky) [inline src/com/owncloud/android/utils/DataStorageUtils.java]: is this a test?
Or do you mean "custom"? If yes, it should be extracted as string ressource.

**Comment by tobiasKaminsky** [comment]
with my motorola x play and an external sd card (used as "internal storage") I only get "external" shown in preferences but cannot change it to the external sd card.
via adb and mount the external sd card is mounted in /mnt/expand/3b0b630d-08a1-4bb1-880a-5993b23097b0

File: src/com/owncloud/android/utils/DataStorageUtils.java (hunk 1 above)
Comment (przybylski) [inline src/com/owncloud/android/utils/DataStorageUtils.java]: It was a testing entry, 'cośtam' means 'something' in polish, I forgot to remove it :)

**Comment by przybylski** [comment]
What's your motorola system version? I am using a trick which is also used in android systems.
But so far there is no support for 6.0. Also 'external' is a naming also used by android API, it might be external device but it can also be internal memory. I think 'default' might be a better wording here.

**Comment by tobiasKaminsky** [comment]
it is android 6 with cyanogen mod

**Comment by przybylski** [comment]
Android 6 changed storage implementation so I need to find a reliable solution for it

**Comment by tobiasKaminsky** [comment]
Please let me know if I can help you (testing, coding or so)

**Comment by przybylski** [comment]
I would like to know if @jancborchardt can switch to his external SD card. But 'costam' is invalid option for now :P

**Comment by tobiasKaminsky** [comment]
pinging @jancborchardt

**Comment by tobiasKaminsky** [comment]
For android 6, with sd card used as "internal storage" this does not work.
There is only "default".
```
shell@lux:/ $ mount
rootfs on / type rootfs (ro,seclabel,relatime)
tmpfs on /dev type tmpfs (rw,seclabel,nosuid,relatime,size=963224k,nr_inodes=113130,mode=755)
devpts on /dev/pts type devpts (rw,seclabel,relatime,mode=600)
proc on /proc type proc (rw,relatime)
sysfs on /sys type sysfs (rw,seclabel,relatime)
selinuxfs on /sys/fs/selinux type selinuxfs (rw,relatime)
debugfs on /sys/kernel/debug type debugfs (rw,relatime)
none on /acct type cgroup (rw,relatime,cpuacct)
none on /sys/fs/cgroup type tmpfs (rw,seclabel,relatime,size=963224k,nr_inodes=113130,mode=750,gid=1000)
tmpfs on /mnt type tmpfs (rw,seclabel,relatime,size=963224k,nr_inodes=113130,mode=755,gid=1000)
none on /dev/cpuctl type cgroup (rw,relatime,cpu)
pstore on /sys/fs/pstore type pstore (rw,relatime)
none on /sys/fs/cgroup/bfqio type cgroup (rw,relatime,bfqio)
/dev/block/bootdevice/by-name/system on /system type ext4 (ro,seclabel,relatime,data=ordered)
/dev/block/bootdevice/by-name/cache on /cache type ext4 (rw,seclabel,nosuid,nodev,noatime,data=ordered)
/dev/block/bootdevice/by-name/fsg on /fsg type ext4 (ro,context=u:object_r:fsg_file:s0,nosuid,nodev,relatime,nobarrier)
/dev/block/bootdevice/by-name/modem on /firmware type ext4 (ro,context=u:object_r:firmware_file:s0,nosuid,nodev,relatime,nobarrier,data=ordered)
/dev/block/bootdevice/by-name/persist on /persist type ext4 (rw,seclabel,nosuid,nodev,noatime,noauto_da_alloc,data=ordered)
adb on /dev/usb-ffs/adb type functionfs (rw,relatime)
tmpfs on /storage type tmpfs (rw,seclabel,relatime,size=963224k,nr_inodes=113130,mode=755,gid=1000)
/dev/block/dm-0 on /data type f2fs (rw,seclabel,nosuid,nodev,noatime,nodiratime,background_gc=on,user_xattr,inline_xattr,acl,inline_data,active_logs=6)
/dev/fuse on /mnt/runtime/default/emulated type fuse (rw,nosuid,nodev,noexec,noatime,user_id=1023,group_id=1023,default_permissions,allow_other)
/dev/fuse on /storage/emulated type fuse (rw,nosuid,nodev,noexec,noatime,user_id=1023,group_id=1023,default_permissions,allow_other)
/dev/fuse on /mnt/runtime/read/emulated type fuse (rw,nosuid,nodev,noexec,noatime,user_id=1023,group_id=1023,default_permissions,allow_other)
/dev/fuse on /mnt/runtime/write/emulated type fuse (rw,nosuid,nodev,noexec,noatime,user_id=1023,group_id=1023,default_permissions,allow_other)
/dev/block/dm-1 on /mnt/expand/a86bc7ac-b798-4d58-b157-d29fae4688ec type ext4 (rw,dirsync,seclabel,nosuid,nodev,noatime)
```

The last entry is my external sd card.

**Comment by tobiasKaminsky** [comment]
When you update this branch can you then do this: #1309 ?
Move storage path to section "details".

**Comment by tobiasKaminsky** [comment]
If I can help you testing: I have a phone with an external sd.

**Comment by tobiasKaminsky** [comment]
@przybylski I have found out that for android 6 you need to use:
context.getExternalFilesDirs(null)
with this you get a File[] Array with all writeable folders: e.g
/storage/emulated/0/Android/data/com.owncloud.android.beta/files
/storage/0F10-100D/Android/data/com.owncloud.android.beta/files

(first one is internal, second is external sd card [mounted as portable storage].

**Comment by tobiasKaminsky** [comment]
Do we check if the external sd card is mounted? https://developer.android.com/guide/topics/data/data-storage.html#MediaAvail

**Comment by tobiasKaminsky** [comment]
I have pushed a suggestion for the lollipop external storage system.
Please adjust it, if you do not like it.
(I will also implement this in beta soon)

**Comment by zoff99** [comment]
https://circleci.com/gh/zoff99/owncloud_android/17#artifacts

**Comment by zoff99** [comment]
needed to do this to compile:

```
## ------- dirty fix ---------
    - find . -type f -name '*.java' 2> /dev/null | while read ff ; do echo "$ff" ; sed -i -e 's#Log_OC.startLogging#// Log_OC.startLogging#g' "$ff" ; done
    - find . -type f -name '*.java' 2> /dev/null | while read ff ; do echo "$ff" ; sed -i -e 's#Log_OC.stopLogging#// Log_OC.stopLogging#g' "$ff" ; done
## ------- dirty fix ---------
```

**Comment by davivel** [comment]
@zoff99 , that's no a good fix, you are just removing the lines that initialize and terminate the logging. The problem is that this branch is not pointing to the commit of the OC Android library it needs.

I'll try something fast to see if it works without removing those lines.

@przybylski , any chance you rebase this PR on master? Thanks in advance.

**Comment by davivel** [comment]
@rhabbachi , that should allow you to build locally the branch, though still needs a real rebase.

**Comment by zoff99** [comment]
for me, i am happy to use the old version, as long as i can change the directory.

**Comment by rickyzhang82** [comment]
Hi Gentlemen,
I love this feature. I'm not an expert on Android. But can you provide a dialog to let user define file path? I believe external SD card location varied in different device and android version.

BTW, don't take it personal. I have trouble to review your PR.

**Comment by jesmrec** [comment]
outdated code, closing... thinking on it for the future

**Pull request closed without merging**